[BUG] Provisioners not migrated to Terraform 0.12

<!--
A great way to contribute to the project is to send a detailed report when you
encounter an issue. We always appreciate a well-written, thorough bug report,
and will thank you for it!
The GitHub issue tracker is for bug reports and feature requests. General support can be found at Slack - gophers.slack.com #terraform channel or message me directly @johandry
-->

**Community Note**

* Please, make sure that we do not have any duplicate issues already open. Search the issue list for this
repository and if there is a duplicate, please close your issue and add a comment to the existing issue instead, you can use the "subscribe" button to get notified on updates
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
* If you have ways to reproduce the issue or have additional information that may help
resolving the issue, please leave a comment.

**Describe the bug**
Terranova works with Terraform 0.12 only if Provisioners are not used. When `AddProvisioner()` is used it fails.

**To Reproduce**
Build the terranova example in `github.com/johandry/terranova-examples/aws/ec2`.

**Expected behavior**
Build successful as well as execution of example in `github.com/johandry/terranova-examples/aws/ec2` or any code using `AddProvisioner()`.

**Screenshots or logs**
```
./main.go:31:42: cannot use file.Provisioner() (type terraform.ResourceProvisioner) as type provisioners.Factory in argument to terranova.NewPlatform(code).AddProvider("aws", "github.com/terraform-providers/terraform-provider-aws/aws".Provider()).AddProvisioner
```

**Platforms**
- `github.com/terraform-providers/terraform-provider-aws v1.60.1-0.20191003145700-f8707a46c6ec`
- `github.com/hashicorp/terraform/builtin/provisioners/file`

**Versions:**
 - Terraformer: [e.g. v0.0.2]
 - Terraform: [e.g. v0.12.9]

**Environment running Terraformer:**
 - OS: macOS

**Additional context:**

**References:**

--- a/platform.go
+++ b/platform.go
@@ -23,6 +23,7 @@ import (
 	"os"
 
 	"github.com/hashicorp/terraform/providers"
+	"github.com/hashicorp/terraform/provisioners"
 	"github.com/hashicorp/terraform/states"
 	"github.com/hashicorp/terraform/states/statefile"
 	"github.com/hashicorp/terraform/terraform"
@@ -33,7 +34,7 @@ import (
 type Platform struct {
 	Code         string
 	Providers    map[string]providers.Factory
-	Provisioners map[string]terraform.ProvisionerFactory
+	Provisioners map[string]provisioners.Factory
 	Vars         map[string]interface{}
 	State        *State
 }
@@ -65,18 +66,13 @@ func (p *Platform) AddProvider(name string, provider terraform.ResourceProvider)
 	return p
 }
 
-func providersFactory(rp terraform.ResourceProvider) providers.Factory {
-	p := NewProvider(rp)
-	return providers.FactoryFixed(p)
-}
-
 func (p *Platform) addDefaultProvisioners() {
-	p.Provisioners = map[string]terraform.ProvisionerFactory{}
+	p.Provisioners = map[string]provisioners.Factory{}
 }
 
 // AddProvisioner adds a new provisioner to the provisioner list
-func (p *Platform) AddProvisioner(name string, provisioner terraform.ProvisionerFactory) *Platform {
-	p.Provisioners[name] = provisioner
+func (p *Platform) AddProvisioner(name string, provisioner terraform.ResourceProvisioner) *Platform {
+	p.Provisioners[name] = provisionersFactory(provisioner)
 	return p
 }
 

--- a/provider.go
+++ b/provider.go
@@ -1,3 +1,19 @@
+/*
+Copyright The Terranova Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
 package terranova
 
 import (
@@ -38,6 +54,11 @@ func NewProvider(provider terraform.ResourceProvider) *Provider {
 	return &Provider{
 		provider: sp,
 	}
+}
+
+func providersFactory(rp terraform.ResourceProvider) providers.Factory {
+	p := NewProvider(rp)
+	return providers.FactoryFixed(p)
 }
 
 // GetSchema implements the GetSchema from providers.Interface. Returns the

--- a/provisioners.go
+++ b/provisioners.go
@@ -1,0 +1,159 @@
+/*
+Copyright The Terranova Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package terranova
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/terraform/configs/configschema"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/provisioners"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/convert"
+)
+
+// Provisioner implements the provisioners.Interface wrapping the legacy
+// terraform.ResourceProvisioner but not using gRPC like terraform does
+type Provisioner struct {
+	provisioner *schema.Provisioner
+	mu          sync.Mutex
+	schema      *configschema.Block
+}
+
+// NewProvisioner creates a Terranova Provisioner to wrap the given legacy ResourceProvisioner
+func NewProvisioner(provisioner terraform.ResourceProvisioner) *Provisioner {
+	sp, ok := provisioner.(*schema.Provisioner)
+	if !ok {
+		return nil
+	}
+	return &Provisioner{
+		provisioner: sp,
+	}
+}
+
+func provisionersFactory(rp terraform.ResourceProvisioner) provisioners.Factory {
+	p := NewProvisioner(rp)
+	return provisioners.FactoryFixed(p)
+}
+
+// GetSchema implements GetSchema from provisioners.Interface. It returns the
+// schema for the provisioner configuration.
+func (p *Provisioner) GetSchema() (resp provisioners.GetSchemaResponse) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.schema != nil {
+		return provisioners.GetSchemaResponse{
+			Provisioner: p.schema,
+		}
+	}
+
+	resp.Provisioner = schema.InternalMap(p.provisioner.Schema).CoreConfigSchema()
+
+	if resp.Provisioner == nil {
+		resp.Diagnostics = resp.Diagnostics.Append(fmt.Errorf("missing provisioner schema"))
+		return resp
+	}
+
+	p.schema = resp.Provisioner
+
+	return resp
+}
+
+// ValidateProvisionerConfig implements ValidateProvisionerConfig from
+// provisioners.Interface. It allows the provisioner to validate the
+// configuration values.
+func (p *Provisioner) ValidateProvisionerConfig(req provisioners.ValidateProvisionerConfigRequest) (resp provisioners.ValidateProvisionerConfigResponse) {
+	cfgSchema := schema.InternalMap(p.provisioner.Schema).CoreConfigSchema()
+	config := terraform.NewResourceConfigShimmed(req.Config, cfgSchema)
+
+	warns, errs := p.provisioner.Validate(config)
+	resp.Diagnostics = appendWarnsAndErrsToDiags(warns, errs)
+
+	return resp
+}
+
+// ProvisionResource implements ProvisionResource from provisioners.Interface.
+// It runs the provisioner with provided configuration.
+// ProvisionResource blocks until the execution is complete.
+// If the returned diagnostics contain any errors, the resource will be
+// left in a tainted state.
+func (p *Provisioner) ProvisionResource(req provisioners.ProvisionResourceRequest) (resp provisioners.ProvisionResourceResponse) {
+	cfgSchema := schema.InternalMap(p.provisioner.Schema).CoreConfigSchema()
+	resourceConfig := terraform.NewResourceConfigShimmed(req.Config, cfgSchema)
+
+	conn := stringMapFromValue(req.Connection)
+
+	instanceState := &terraform.InstanceState{
+		Ephemeral: terraform.EphemeralState{
+			ConnInfo: conn,
+		},
+		Meta: make(map[string]interface{}),
+	}
+
+	if err := p.provisioner.Apply(req.UIOutput, instanceState, resourceConfig); err != nil {
+		resp.Diagnostics = resp.Diagnostics.Append(err)
+	}
+
+	return resp
+}
+
+// Stop implements Stop from provisioners.Interface. It is called to interrupt
+// the provisioner.
+//
+// Stop should not block waiting for in-flight actions to complete. It
+// should take any action it wants and return immediately acknowledging it
+// has received the stop request. Terraform will not make any further API
+// calls to the provisioner after Stop is called.
+//
+// The error returned, if non-nil, is assumed to mean that signaling the
+// stop somehow failed and that the user should expect potentially waiting
+// a longer period of time.
+func (p *Provisioner) Stop() error {
+	return p.provisioner.Stop()
+}
+
+// Close implements Close from provisioners.Interface. It shuts down the plugin
+// process if applicable.
+func (p *Provisioner) Close() error {
+	return nil
+}
+
+// stringMapFromValue converts a cty.Value to a map[stirng]string.
+// This will panic if the val is not a cty.Map(cty.String).
+func stringMapFromValue(val cty.Value) map[string]string {
+	m := map[string]string{}
+	if val.IsNull() || !val.IsKnown() {
+		return m
+	}
+
+	for it := val.ElementIterator(); it.Next(); {
+		ak, av := it.Element()
+		name := ak.AsString()
+
+		if !av.IsKnown() || av.IsNull() {
+			continue
+		}
+
+		av, _ = convert.Convert(av, cty.String)
+		m[name] = av.AsString()
+	}
+
+	return m
+}

